File: app/src/main/java/com/georgcantor/wallpaperapp/ui/PicDetailActivity.java
package com.georgcantor.wallpaperapp.ui;

import android.Manifest;
import android.app.DownloadManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.AndroidRuntimeException;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.georgcantor.wallpaperapp.R;
import com.georgcantor.wallpaperapp.model.Hit;
import com.georgcantor.wallpaperapp.model.db.DatabaseHelper;
import com.georgcantor.wallpaperapp.network.NetworkUtilities;
import com.georgcantor.wallpaperapp.ui.adapter.TagAdapter;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import jp.wasabeef.picasso.transformations.CropCircleTransformation;

public class PicDetailActivity extends AppCompatActivity {

    public static final String EXTRA_PIC = "picture";
    public static final String ORIGIN = "caller";
    private Hit hit;
    private List<String> tags = new ArrayList<>();
    int first = 0;
    public NetworkUtilities networkUtilities;
    public RecyclerView recyclerView;
    public TagAdapter tagAdapter;
    private File file;
    private TextView tagTitle;
    private int permissionCheck1;
    private ProgressBar progressBar;
    private FloatingActionButton fab;
    private DatabaseHelper db;
    public boolean isDownloaded = false;
    public boolean isCallerCollection = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        networkUtilities = new NetworkUtilities(this);
        setContentView(R.layout.activity_pic_detail);
        fab = findViewById(R.id.fab_download);

        progressBar = findViewById(R.id.progressBarDetail);
        progressBar.setVisibility(View.VISIBLE);
        db = new DatabaseHelper(this);

        tagTitle = findViewById(R.id.toolbar_title);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        initView();

        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!fileIsExist()) {
                    if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        checkPermisson();
                        if (permissionCheck1 != PackageManager.PERMISSION_GRANTED) {
                            return;
                        }
                    }
                    if (networkUtilities.isInternetConnectionPresent()) {
                        AlertDialog.Builder builder =
                                new AlertDialog.Builder(PicDetailActivity.this);
                        builder.setTitle(R.string.download);
                        builder.setIcon(R.drawable.ic_download);
                        builder.setMessage(R.string.choose_format);

                        builder.setPositiveButton("HD", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                progressBar.setVisibility(View.VISIBLE);
                                if (permissionCheck1 == PackageManager.PERMISSION_GRANTED) {
                                    if (!fileIsExist()) {
                                        String uri = hit.getWebformatURL();
                                        Uri image_uri = Uri.parse(uri);
                                        downloadData(image_uri);
                                        fab.setImageDrawable(getApplicationContext().getResources()
                                                .getDrawable(R.drawable.ic_photo));
                                    } else {
                                        Toast.makeText(PicDetailActivity.this, getResources()
                                                        .getString(R.string.image_downloaded),
                                                Toast.LENGTH_SHORT).show();
                                        progressBar.setVisibility(View.GONE);
                                    }
                                }
                            }
                        });

                        builder.setNeutralButton(hit.getImageWidth() + " x "
                                + hit.getImageHeight(), new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                progressBar.setVisibility(View.VISIBLE);
                                if (permissionCheck1 == PackageManager.PERMISSION_GRANTED) {
                                    if (!fileIsExist()) {
                                        String uri = hit.getImageURL();
                                        Uri image_uri = Uri.parse(uri);
                                        downloadData(image_uri);
                                        fab.setImageDrawable(getApplicationContext().getResources()
                                                .getDrawable(R.drawable.ic_photo));
                                    } else {
                                        Toast.makeText(PicDetailActivity.this, getResources()
                                                        .getString(R.string.image_downloaded),
                                                Toast.LENGTH_SHORT).show();
                                        progressBar.setVisibility(View.GONE);
                                    }
                                }
                            }
                        });

                        builder.setNegativeButton("FullHD", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                progressBar.setVisibility(View.VISIBLE);
                                if (permissionCheck1 == PackageManager.PERMISSION_GRANTED) {
                                    if (!fileIsExist()) {
                                        String uri = hit.getFullHDURL();
                                        Uri image_uri = Uri.parse(uri);
                                        downloadData(image_uri);
                                        fab.setImageDrawable(getApplicationContext().getResources()
                                                .getDrawable(R.drawable.ic_photo));
                                    } else {
                                        Toast.makeText(PicDetailActivity.this, getResources()
                                                        .getString(R.string.image_downloaded),
                                                Toast.LENGTH_SHORT).show();
                                        progressBar.setVisibility(View.GONE);
                                    }
                                }
                            }
                        });
                        builder.create().show();
                    } else {
                        Toast.makeText(PicDetailActivity.this, getResources()
                                .getString(R.string.no_internet), Toast.LENGTH_SHORT).show();
                    }

                } else {
                    Uri sendUri2 = Uri.fromFile(file);
                    Log.d(getResources().getString(R.string.URI), sendUri2.toString());
                    Intent intent = new Intent(Intent.ACTION_ATTACH_DATA);
                    intent.setDataAndType(sendUri2, getResources().getString(R.string.image_jpg));
                    intent.putExtra(getResources().getString(R.string.mimeType),
                            getResources().getString(R.string.image_jpg));
                    startActivityForResult(Intent.createChooser(intent,
                            getResources().getString(R.string.Set_As)), 200);
                }
            }
        });
    }

    private void initView() {
        permissionCheck1 = ContextCompat.checkSelfPermission(this,
                Manifest.permission.WRITE_EXTERNAL_STORAGE);

        if (getIntent().hasExtra(EXTRA_PIC)) {
            hit = getIntent().getParcelableExtra(EXTRA_PIC);
        } else {
            throw new IllegalArgumentException("Detail activity must receive a Hit parcelable");
        }
        String title = hit.getTags();
        while (title.contains(",")) {
            String f = title.substring(0, title.indexOf(","));
            tags.add(f);
            first = title.indexOf(",");
            title = title.substring(++first);
        }
        tags.add(title);
        tagTitle.setText(tags.get(0));
        ImageView wallp = findViewById(R.id.wallpaper_detail);
        TextView fav = findViewById(R.id.fav);
        TextView userId = findViewById(R.id.user_name);
        ImageView userImage = findViewById(R.id.user_image);
        TextView downloads = findViewById(R.id.down);
        recyclerView = findViewById(R.id.tagsRv);
        recyclerView.setLayoutManager(new LinearLayoutManager(this,
                LinearLayoutManager.HORIZONTAL, false));
        tagAdapter = new TagAdapter(this);
        tagAdapter.setTagList(tags);
        recyclerView.setAdapter(tagAdapter);
        file = new File(Environment.getExternalStoragePublicDirectory("/"
                + getResources().getString(R.string.app_name)), hit.getId()
                + getResources().getString(R.string.jpg));
        if (fileIsExist()) {
            fab.setImageDrawable(getApplicationContext().getResources()
                    .getDrawable(R.drawable.ic_photo));
        }

        if (getIntent().hasExtra(ORIGIN)) {
            Picasso.with(this)
                    .load(file)
                    .placeholder(R.drawable.plh)
                    .into(wallp);
            isCallerCollection = true;
        } else {
            Picasso.with(this)
                    .load(hit.getWebformatURL())
                    .placeholder(R.drawable.plh)
                    .into(wallp, new Callback() {
                        @Override
                        public void onSuccess() {
                            progressBar.setVisibility(View.GONE);
                        }

                        @Override
                        public void onError() {
                            progressBar.setVisibility(View.GONE);
                            Toast.makeText(PicDetailActivity.this,
                                    getString(R.string.something_went_wrong),
                                    Toast.LENGTH_SHORT).show();
                        }
                    });
        }

        userId.setText(hit.getUser());
        downloads.setText(String.valueOf(hit.getDownloads()));
        fav.setText(String.valueOf(hit.getFavorites()));
        if (!networkUtilities.isInternetConnectionPresent()) {
            Picasso.with(this)
                    .load(R.drawable.memb)
                    .transform(new CropCircleTransformation())
                    .into(userImage);
        } else {
            if (!hit.getUserImageURL().isEmpty()) {
                Picasso.with(this)
                        .load(hit.getUserImageURL())
                        .transform(new CropCircleTransformation())
                        .into(userImage);
            } else {
                Picasso.with(this)
                        .load(R.drawable.memb)
                        .transform(new CropCircleTransformation())
                        .into(userImage);
            }
        }
        IntentFilter filter = new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE);
        registerReceiver(downloadReceiver, filter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_fav, menu);
        if (db.containFav(String.valueOf(hit.getPreviewURL()))) {
            menu.findItem(R.id.action_add_to_fav).setIcon(R.drawable.ic_star_red_24dp);
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                break;
            case R.id.action_add_to_fav:
                if (!db.containFav(String.valueOf(hit.getPreviewURL()))) {
                    addToFavorite(String.valueOf(hit.getPreviewURL()));
                    item.setIcon(R.drawable.ic_star_red_24dp);
                    Toast.makeText(this, R.string.add_to_fav_toast, Toast.LENGTH_SHORT).show();
                } else {
                    db.deleteFromFavorites(String.valueOf(hit.getPreviewURL()));
                    item.setIcon(R.drawable.ic_star_border_black_24dp);
                    Toast.makeText(this, R.string.del_from_fav_toast, Toast.LENGTH_SHORT).show();
                }
                break;
            case R.id.action_share:
                try {
                    Intent i = new Intent(Intent.ACTION_SEND);
                    i.setType("text/plain");
                    i.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.app_name));
                    String sAux = hit.getImageURL();
                    i.putExtra(Intent.EXTRA_TEXT, sAux);
                    startActivity(Intent.createChooser(i, getString(R.string.choose_share)));
                } catch (AndroidRuntimeException e) {
                    e.printStackTrace();
                    Toast.makeText(this, "Can not share image", Toast.LENGTH_SHORT).show();
                }
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private long downloadData(Uri uri) {
        long downloadReference;
        DownloadManager downloadManager = (DownloadManager) getSystemService(DOWNLOAD_SERVICE);

        String name = Environment.getExternalStorageDirectory().getAbsolutePath();
        name += "/YourDirectoryName/";

        DownloadManager.Request request = new DownloadManager.Request(uri);
        request.setTitle(tags.get(0) + getResources().getString(R.string.down));
        request.setDescription(getResources().getString(R.string.down_wallpapers));
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            request.setDestinationInExternalPublicDir("/"
                    + getResources().getString(R.string.app_name), hit.getId()
                    + getResources().getString(R.string.jpg));
        }
        downloadReference = downloadManager.enqueue(request);

        return downloadReference;
    }

    private BroadcastReceiver downloadReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Toast toast = Toast.makeText(context, tags.get(0) + getResources()
                    .getString(R.string.down_complete), Toast.LENGTH_SHORT);
            toast.show();
            isDownloaded = true;
            progressBar.setVisibility(View.GONE);
        }
    };

    @Override
    public void onDestroy() {
        try {
            if (downloadReceiver != null)
                unregisterReceiver(downloadReceiver);
        } catch (Exception e) {
            e.printStackTrace();
        }
        super.onDestroy();
    }

    public boolean fileIsExist() {
        return file.exists();
    }

    public void checkPermisson() {
        if (permissionCheck1 != PackageManager.PERMISSION_GRANTED) {
            int MY_PERMISSIONS_REQUEST_WRITE_STORAGE = 102;
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission
                    .WRITE_EXTERNAL_STORAGE}, MY_PERMISSIONS_REQUEST_WRITE_STORAGE);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            restartActivity();
        } else {
            Intent intent = new Intent();
            intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
            Uri uri = Uri.fromParts("package", getPackageName(), null);
            intent.setData(uri);
            startActivity(intent);
            finish();
            Toast.makeText(this, R.string.you_need_perm_toast, Toast.LENGTH_LONG).show();
        }
    }

    public void restartActivity() {
        hit = getIntent().getParcelableExtra(EXTRA_PIC);
        Intent intent = new Intent(PicDetailActivity.this,
                PicDetailActivity.class);
        intent.putExtra(EXTRA_PIC, hit);
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
    }

    public void addToFavorite(String imageUrl) {
        db.insertToFavorites(imageUrl);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        overridePendingTransition(R.anim.pull_in_left, R.anim.push_out_right);
    }
}

File: app/src/main/java/com/georgcantor/wallpaperapp/ui/PicDetailActivity.kt
package com.georgcantor.wallpaperapp.ui

import android.Manifest
import android.app.DownloadManager
import android.app.WallpaperManager
import android.content.*
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.provider.Settings
import android.util.AndroidRuntimeException
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.georgcantor.wallpaperapp.R
import com.georgcantor.wallpaperapp.model.CommonPic
import com.georgcantor.wallpaperapp.model.local.db.DatabaseHelper
import com.georgcantor.wallpaperapp.ui.adapter.TagAdapter
import com.georgcantor.wallpaperapp.ui.util.*
import com.google.gson.Gson
import com.squareup.picasso.Callback
import com.squareup.picasso.Picasso
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import jp.wasabeef.picasso.transformations.CropCircleTransformation
import kotlinx.android.synthetic.main.activity_detail.detailImageView
import kotlinx.android.synthetic.main.activity_detail.downloadAnimationView
import kotlinx.android.synthetic.main.activity_detail.downloadsTextView
import kotlinx.android.synthetic.main.activity_detail.fabSetWall
import kotlinx.android.synthetic.main.activity_detail.favoritesTextView
import kotlinx.android.synthetic.main.activity_detail.nameTextView
import kotlinx.android.synthetic.main.activity_detail.progressAnimationView
import kotlinx.android.synthetic.main.activity_detail.tagsRecyclerView
import kotlinx.android.synthetic.main.activity_detail.userImageView
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.IOException
import java.util.*

class PicDetailActivity : AppCompatActivity() {

    companion object {
        const val EXTRA_PIC = "picture"
        const val PREF_BOOLEAN = "is_set_wall"
        const val MY_PREFS = "my_prefs"
    }

    private var hit: CommonPic? = null
    private val tags = ArrayList<String>()
    private var first = 0
    private lateinit var tagAdapter: TagAdapter
    private var file: File? = null
    private var tagTitle: TextView? = null
    private var permissionCheck: Int = 0
    private var db: DatabaseHelper? = null
    private var pathOfFile: String? = null
    private lateinit var prefs: SharedPreferences
    private lateinit var editor: SharedPreferences.Editor

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        setContentView(R.layout.activity_detail)
        progressAnimationView?.showAnimation()

        db = DatabaseHelper(this)

        editor = getSharedPreferences(MY_PREFS, Context.MODE_PRIVATE).edit()
        prefs = getSharedPreferences(MY_PREFS, Context.MODE_PRIVATE)

        supportActionBar?.setHomeAsUpIndicator(R.drawable.ic_arrow_back)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setBackgroundDrawable(ColorDrawable(ContextCompat.getColor(this, R.color.colorPrimary)))

        initView()

        fabSetWall.setOnClickListener {
            if (this.isNetworkAvailable()) {
                if (permissionCheck == PackageManager.PERMISSION_GRANTED) {
                    val uri = Uri.fromFile(file)
                    pathOfFile = UtilityMethods.getPath(applicationContext, uri)
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        progressAnimationView?.showAnimation()
                        setWallAsync()
                    } else {
                        setAsWallpaper()
                    }
                } else {
                    editor.putBoolean(PREF_BOOLEAN, true)
                    editor.apply()
                    checkSavingPermission()
                }
            } else {
                longToast(getString(R.string.no_internet))
            }
        }
    }

    private fun setAsWallpaper() {
        val uri = Uri.fromFile(file)
        val intent = Intent(Intent.ACTION_ATTACH_DATA)
        intent.setDataAndType(uri, getString(R.string.image_jpg))
        intent.putExtra(getString(R.string.mimeType), getString(R.string.image_jpg))

        startActivityForResult(
            Intent.createChooser(intent, getString(R.string.Set_As)), 200
        )
    }

    private val downloadReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            try {
                shortToast(tags[0] + getString(R.string.down_complete))
                downloadAnimationView?.hideAnimation()
            } catch (e: IndexOutOfBoundsException) {
                downloadAnimationView?.hideAnimation()
            }
        }
    }

    @RequiresApi(Build.VERSION_CODES.KITKAT)
    private fun setWallAsync() {
        progressAnimationView?.showAnimation()

        val disposable = getBitmapAsync()?.subscribe({
            val wallpaperManager = WallpaperManager.getInstance(baseContext)
            it?.let { bitmap ->
                getImageUri(bitmap, applicationContext).subscribe({ uri ->
                    try {
                        startActivity(
                            Intent(
                                wallpaperManager.getCropAndSetWallpaperIntent(
                                    uri
                                )
                            )
                        )
                    } catch (e: IllegalArgumentException) {
                        try {
                            it.let { bitMap ->
                                getImageUri(bitMap, applicationContext).subscribe({ uri ->
                                    val bitmap2 = MediaStore.Images.Media.getBitmap(
                                        contentResolver,
                                        uri
                                    )
                                    WallpaperManager.getInstance(this@PicDetailActivity)
                                        .setBitmap(bitmap2)
                                }, {
                                    shortToast(getString(R.string.something_went_wrong))
                                })
                            }
                        } catch (e: OutOfMemoryError) {
                            shortToast(getString(R.string.something_went_wrong))
                        }
                    }
                }, { throwable ->
                    longToast(throwable.message.toString())
                })
            }
            shortToast(getString(R.string.wallpaper_is_install))
            recreate()
        }, {
            longToast(getString(R.string.something_went_wrong))
        })

        disposable?.let(DisposableManager::add)
    }

    private fun getBitmapAsync(): Observable<Bitmap?>? {
        return Observable.fromCallable {
            var result: Bitmap? = null
            try {
                result = Picasso.with(applicationContext)
                    .load(hit?.imageURL)
                    .get()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            result
        }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
    }

    private fun getImageUri(inImage: Bitmap, inContext: Context): Observable<Uri> {
        return Observable.fromCallable {
            val bytes = ByteArrayOutputStream()
            inImage.compress(Bitmap.CompressFormat.JPEG, 100, bytes)
            val path = MediaStore.Images.Media.insertImage(
                inContext.contentResolver,
                inImage, "Title", null
            )
            Uri.parse(path)
        }
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
    }

    private fun initView() {
        permissionCheck = ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
        )

        if (intent.hasExtra(EXTRA_PIC)) {
            hit = intent.getParcelableExtra(EXTRA_PIC)
        } else {
            shortToast(getString(R.string.something_went_wrong))
        }
        hit?.let {
            var title = it.tags
            while (title?.contains(",") == true) {
                val element = title.substring(0, title.indexOf(","))
                tags.add(element)
                first = title.indexOf(",")
                title = title.substring(++first)
            }
            title?.let(tags::add)
        }

        tagTitle?.text = tags[0]
        tagsRecyclerView.layoutManager = LinearLayoutManager(
            this,
            LinearLayoutManager.HORIZONTAL, false
        )
        tagAdapter = TagAdapter(this)
        tagAdapter.setTagList(tags)
        tagsRecyclerView.adapter = tagAdapter
        file = File(
            Environment.getExternalStoragePublicDirectory(
                "/" + resources
                    .getString(R.string.app_name)
            ), hit?.id.toString() + resources
                .getString(R.string.jpg)
        )

        Picasso.with(this)
            .load(hit?.fullHDURL)
            .placeholder(R.drawable.plh)
            .into(detailImageView, object : Callback {
                override fun onSuccess() {
                    progressAnimationView?.hideAnimation()
                }

                override fun onError() {
                    progressAnimationView?.hideAnimation()
                    shortToast(getString(R.string.something_went_wrong))
                }
            })

        nameTextView.text = hit?.user
        downloadsTextView.text = hit?.downloads.toString()
        favoritesTextView.text = hit?.favorites.toString()
        if (!this.isNetworkAvailable()) {
            Picasso.with(this)
                .load(R.drawable.memb)
                .transform(CropCircleTransformation())
                .into(userImageView)
        } else {
            hit?.let {
                if (it.userImageURL?.isNotEmpty() == true) {
                    Picasso.with(this)
                        .load(hit?.userImageURL)
                        .transform(CropCircleTransformation())
                        .into(userImageView)
                } else {
                    Picasso.with(this)
                        .load(R.drawable.memb)
                        .transform(CropCircleTransformation())
                        .into(userImageView)
                }
            }
        }
        val filter = IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE)
        registerReceiver(downloadReceiver, filter)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_fav, menu)
        db?.let {
            if (it.containFav(hit?.url.toString())) {
                menu.findItem(R.id.action_add_to_fav).setIcon(R.drawable.ic_star_red_24dp)
            }
        }

        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        db?.let { db ->
            when (item.itemId) {
                android.R.id.home -> onBackPressed()
                R.id.action_add_to_fav -> if (!db.containFav(hit?.url.toString())) {
                    hit?.let { addToFavorite(it.url.toString(), it.imageURL.toString(), it) }
                    item.setIcon(R.drawable.ic_star_red_24dp)
                    shortToast(getString(R.string.add_to_fav_toast))
                } else {
                    db.deleteFromFavorites(hit?.url.toString())
                    item.setIcon(R.drawable.ic_star_border)
                    shortToast(getString(R.string.del_from_fav_toast))
                }
                R.id.action_share -> try {
                    val intent = Intent(Intent.ACTION_SEND)
                    intent.type = "text/plain"
                    intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.app_name))
                    val sAux = hit?.url
                    intent.putExtra(Intent.EXTRA_TEXT, sAux)
                    startActivity(Intent.createChooser(intent, getString(R.string.choose_share)))
                } catch (e: AndroidRuntimeException) {
                    e.printStackTrace()
                    shortToast(getString(R.string.cant_share))
                }
                R.id.action_download -> {
                    if (this.isNetworkAvailable()) {
                        if (permissionCheck == PackageManager.PERMISSION_GRANTED) {
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                                downloadPictureQ(hit?.url ?: "")
                            } else {
                                downloadPicture()
                            }
                        } else {
                            checkSavingPermission()
                        }
                    } else {
                        shortToast(getString(R.string.no_internet))
                    }
                }
                else -> {
                }
            }
        }

        return super.onOptionsItemSelected(item)
    }

    private fun downloadPicture(): Long {
        downloadAnimationView?.showAnimation()
        val uri = hit?.imageURL
        val imageUri = Uri.parse(uri)

        val downloadReference: Long
        val downloadManager =
            getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager

        var name = Environment.getExternalStorageDirectory().absolutePath
        name += "/YourDirectoryName/"

        val request = DownloadManager.Request(imageUri)

        try {
            request.setTitle(tags[0] + getString(R.string.down))
            request.setDescription(getString(R.string.down_wallpapers))
            if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
                request.setDestinationInExternalPublicDir(
                    "/" + resources
                        .getString(R.string.app_name), hit?.id.toString() + resources
                        .getString(R.string.jpg)
                )
            }
        } catch (e: IllegalStateException) {
            shortToast(getString(R.string.something_went_wrong))
        } catch (e: IndexOutOfBoundsException) {
            shortToast(getString(R.string.something_went_wrong))
        }
        downloadReference = downloadManager.enqueue(request)

        return downloadReference
    }

    private fun downloadPictureQ(url: String) {
        downloadAnimationView?.visibility = View.VISIBLE
        downloadAnimationView?.playAnimation()
        downloadAnimationView?.loop(true)

        val name = url.getImageNameFromUrl()

        val downloadManager = getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager?
        val request = DownloadManager.Request(Uri.parse(url))

        request.setAllowedNetworkTypes(DownloadManager.Request.NETWORK_WIFI or DownloadManager.Request.NETWORK_MOBILE)
            .setAllowedOverRoaming(false)
            .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE)
            .setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, name)

        downloadManager?.enqueue(request)
    }

    private fun checkSavingPermission() {
        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
            val requestCode = 102
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
                requestCode
            )
        }
    }

    @RequiresApi(Build.VERSION_CODES.KITKAT)
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            val isSetWall = prefs.getBoolean(PREF_BOOLEAN, false)
            if (isSetWall) setWallAsync() else downloadPicture()
        } else {
            val intent = Intent()
            intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
            val uri = Uri.fromParts("package", packageName, null)
            intent.data = uri
            startActivity(intent)
            finish()
            longToast(getString(R.string.you_need_perm_toast))
        }
    }

    private fun addToFavorite(imageUrl: String, hdUrl: String, commonPic: CommonPic) {
        val gson = Gson()
        val toStoreObject = gson.toJson(commonPic)
        db?.insertToFavorites(imageUrl, hdUrl, toStoreObject)
    }

    override fun onBackPressed() {
        super.onBackPressed()
        overridePendingTransition(R.anim.pull_in_left, R.anim.push_out_right)
    }

    public override fun onDestroy() {
        try {
            unregisterReceiver(downloadReceiver)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        DisposableManager.dispose()
        super.onDestroy()
    }

}
